recastjs/src/recastjs.cpp

```diff
@@ -0,0 +1,577 @@
+#include "recastjs.h"
+#include "Recast.h"
+#include "DetourNavMesh.h"
+#include "DetourNavMeshBuilder.h"
+#include "DetourNavMesh.h"
+#include "DetourNavMeshQuery.h"
+
+#include <stdio.h>
+#include <vector>
+#include <float.h>
+#include <algorithm>
+#include <math.h>
+#include <sstream>
+#include <iostream>
+
+void Log(const char* str)
+{
+    std::cout << std::string(str) << std::endl;
+}
+
+int g_seed = 1337;
+inline int fastrand() 
+{ 
+    g_seed = (214013*g_seed+2531011); 
+    return (g_seed>>16)&0x7FFF; 
+} 
+
+inline float r01()
+{
+    return ((float)fastrand())*(1.f/32767.f);
+}
+
+void NavMesh::destroy()
+{
+    dtFreeNavMesh(m_navMesh);
+    dtFreeNavMeshQuery(m_navQuery);
+}
+
+void NavMesh::build(const float* positions, const int positionCount, const int* indices, const int indexCount, const rcConfig& config)
+{
+    std::vector<Vec3> mTriangles;
+    const float* pv = &positions[0];
+    const int* t = &indices[0];
+
+    Vec3 bbMin(FLT_MAX);
+    Vec3 bbMax(-FLT_MAX);
+    mTriangles.resize(indexCount);
+    for (unsigned int i = 0; i<indexCount; i++)
+    {
+        int ind = (*t++) * 3;
+        Vec3 v( pv[ind+2], pv[ind+1], pv[ind] );
+        bbMin.isMinOf( v );
+        bbMax.isMaxOf( v );
+        mTriangles[i] = v;
+    }
+
+    bool m_keepInterResults = false;
+
+
+    // Set the area where the navigation will be build.
+    // Here the bounds of the input mesh are used, but the
+    // area could be specified by an user defined box, etc.
+    //float bmin[3] = {-20.f, 0.f, -20.f};
+    //float bmax[3] = { 20.f, 1.f,  20.f};
+    rcConfig m_cfg = config;
+    m_cfg.walkableHeight = (int)ceilf(config.walkableHeight / m_cfg.ch);
+    m_cfg.walkableClimb = (int)floorf(config.walkableClimb / m_cfg.ch);
+    m_cfg.walkableRadius = (int)ceilf(config.walkableRadius / m_cfg.cs);
+    m_cfg.maxEdgeLen = (int)(config.maxEdgeLen / config.cs);
+    m_cfg.maxSimplificationError = config.maxSimplificationError;
+    m_cfg.minRegionArea = (int)rcSqr(config.minRegionArea);        // Note: area = size*size
+    m_cfg.mergeRegionArea = (int)rcSqr(config.mergeRegionArea);    // Note: area = size*size
+    m_cfg.maxVertsPerPoly = (int)config.maxVertsPerPoly;
+    m_cfg.detailSampleDist = config.detailSampleDist < 0.9f ? 0 : config.cs * config.detailSampleDist;
+    m_cfg.detailSampleMaxError = config.ch * config.detailSampleMaxError;
+
+    rcVcopy(m_cfg.bmin, &bbMin.x);
+    rcVcopy(m_cfg.bmax, &bbMax.x);
+    
+
+    rcCalcGridSize(m_cfg.bmin, m_cfg.bmax, m_cfg.cs, &m_cfg.width, &m_cfg.height);
+
+    rcContext ctx;
+    rcContext* m_ctx = &ctx;
+
+    //
+    // Step 2. Rasterize input polygon soup.
+    //
+    rcHeightfield* m_solid;
+    // Allocate voxel heightfield where we rasterize our input data to.
+    m_solid = rcAllocHeightfield();
```

Comment:
Seems like this is leaked in a bunch of places below where errors happen? Perhaps it should be wrapped in a RAII? If you pull in GSL, you can use GSL::finally to do it.